### ISSUE TITLE
New buzzwords Poodle Pug or Wiener Dog? Deploy a Dog Identification TensorFlow Model Using Python and Flask

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#2359]
Poodle, Pug, or Wiener Dog? Deploy a Dog Identification TensorFlow Model Using Python and Flask<br>
http://ift.tt/2zguXFZ<br>
<img width="640" height="426" src="http://ift.tt/2kIcmOo" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2017/12/8fc12a14-dogs-2200676_640.jpg">
<p>In this post, we&rsquo;ll create a demo to see how simple it is to develop a <a href="http://ift.tt/2eQa1gK" class="local-link">machine learning</a>-based service using Python&rsquo;s Flask library.</p>
<p>There are a number of cases where you might not be able to use a cloud service to host your model and would be required to roll-your-own inference service. In many large enterprises, on-premise solutions are mandatory. Approvals to purchase third-party solutions can also be lengthy and complex. So developing your own small service may be the best solution.</p>
<p>The demo we will develop here will be a proof-of-concept demonstration of how to load and host your model but won&rsquo;t go into issues around security, authentication or load balancing.</p>
<h2>Developing the Model</h2>
<div class="clearfix infoBlock" target="_blank">
<div class="infoBlockImageBlock"><img class="infoBlockImage" src="http://ift.tt/2kETfoE" style="border-radius: 50%;"></div>
<div class="infoBlockTextBlock">
<div class="infoBlockTitle">Pete Garcin</div>
<div class="infoBlockText">As Developer Advocate at ActiveState, Pete has over 15 years in software development in both web and games having shipped over 40 titles in roles ranging from Programmer to Audio Director to Executive Producer. He earned his undergraduate degree at University of Waterloo, and an MA in Communication from Carleton University in Ottawa. He&rsquo;s passionate about engaging with communities and dedicated to enhancing developers&rsquo; experiences with ActiveState products.</div>
</div>
</div>
<p>The first stage is to develop a model. You may already have a model that you&rsquo;ve been training that&rsquo;s ready to be deployed. There are a few ways to ready your model for deployment using <a href="http://ift.tt/1Xwlwg0" class="ext-link" rel="external nofollow">TensorFlow</a>. For instance, you can use TensorFlow&rsquo;s &lsquo;freeze graph&rsquo; functionality to export the entire graph as a Protocol Buffer to use for serving.</p>
<p>For our demo, we&rsquo;ll develop a super-powerful AI that is able to determine the difference between poodles, pugs, and &ldquo;Wiener Dogs&rdquo; (aka dachshunds). For this bleeding edge technology, we can use a retrained version of Google&rsquo;s <a href="http://ift.tt/1MjKVDR" class="ext-link" rel="external nofollow">Inception</a> or <a href="http://ift.tt/2roMIPB" class="ext-link" rel="external nofollow">MobileNet</a> models.</p>
<p>Now, much like the famous <a href="http://ift.tt/2uaC5RY" class="ext-link" rel="external nofollow">Silicon Valley &ldquo;Not Hotdog&rdquo;</a> app, our <em>actual dog</em> identification algorithm will surely be of great utility to the human race. So once we&rsquo;ve trained our model to identify these key dog breeds we&rsquo;ll want to create a service to make it available to the world.</p>
<p>Retraining the Inception or MobileNet models is actually extremely straightforward. You can train it to classify entirely new categories of objects simply by providing a series of images in a labeled subfolder. It will then take those images, classify them with the label (ie. the folder name) and output a newly trained model in .pb format.</p>
<p><strong>TIP: If you&rsquo;re looking for an excellent introduction and tutorial on how to retrain these models, you can work through this <a href="http://ift.tt/2flXYrT" class="ext-link" rel="external nofollow">codelab</a></strong><strong>&nbsp;provided by Google.</strong></p>
<p>For performance reasons, we&rsquo;re choosing the MobileNet model. It is optimized for use on mobile devices. This model sacrifices some accuracy for speed but works pretty well for our use case.</p>
<p>We can then adapt one of Google&rsquo;s label scripts that contains useful utility functions for loading graphs and transforming images into tensors to use with TensorFlow. This adapted script will become the basis for our inference service.</p>
<h2>Building the Service</h2>
<p>Once we&rsquo;ve retrained our model and exported it to disk we can host the model as a service. We&rsquo;ll load the model from disk with a simple function that takes the graph definition directly from the file and uses that to generate a graph. TensorFlow does most of this for us:</p>
<pre>
def load_graph(<em>model_file</em>):
 graph = tf.Graph()
 graph_def = tf.GraphDef()

 with open(model_file, "rb") as f:
   graph_def.ParseFromString(f.read())
 with graph.as_default():
   tf.import_graph_def(graph_def)
</pre>
<p>Using the Python <a href="http://flask.pocoo.org/" class="ext-link" rel="external nofollow">Flask</a> microframework, much of the heavy-lifting around configuring a server and handling requests is done for us. After we&rsquo;ve created a Flask app object:</p>
<pre>
app = Flask(__name__)
</pre>
<p>Then we can easily create routes for where our classification service will live. Let&rsquo;s create a default route to our classify() function that will allow us to pass an image to the endpoint for identification.</p>
<pre>
@app.route('/')

def classify():
</pre>
<p>Using the decorator syntax to define the route, it will configure the service so that our <strong>classify()</strong> function will be called every time someone hits the root of our service address. We said we wanted users to be able to specify a file to be identified so we&rsquo;ll store that as a parameter from the request:</p>
<pre>
file_name = request.args['file']
</pre>
<p>In an actual app, we&rsquo;d probably populate this from a form attachment or URL. For our example, we&rsquo;ll simply let users specify a path to the file that they want to be identified.</p>
<p>We can then read the image file and turn it into a tensor to be used as input to the graph we loaded previously. The base script included a number of useful functions including <strong>read_tensor_from_image_file()</strong> which will take the image file and turn it into a tensor to use as input by using a small custom TensorFlow graph.</p>
<p>Running the inference on our graph with this image is again quite straightforward:</p>
<pre>
       results = sess.run(output_operation.outputs[0],
                     {input_operation.outputs[0]: t})
</pre>
<p>In this line, the variable &lsquo;t&rsquo; represents the image tensor that was created by read_tensor_from_image_file() function. TensorFlow will then take that image and run the new retrained model to generate predictions.</p>
<p>Those predictions come as a series of probabilities that indicate which of the classes (poodle, pug, or wiener dog) is the most likely. Since this is just a prediction service, it will simply return a JSON representation of the arrays.</p>
<p>Inside our script we can start our service with:</p>
<pre>
app.run(<em>debug</em>=True, <em>port</em>=8000)
</pre>
<p>Then, if we want to launch the script from the command line, all we have to do is run `python app.py` and it will initialize and start running on port 8000.</p>
<h2>Using the Service</h2>
<p>We can now use this service either by visiting it in a web browser or generally making any REST call on that port. For an easy test we can access it using `curl`:</p>
<pre>
curl localhost:8000?file=path/to/poodle.jpg
</pre>
<p>So, if we feed it the following image of my standard poodle, Orion, let&rsquo;s see what it says:</p>
<p><img class="aligncenter size-full wp-image-3722577" src="http://ift.tt/2l19KuD" width="975" height="650" data-id="3722577"></p>
<p>The output is:</p>
<pre>
[
  [
    "poodle",
    "pug",
    "dachshund"
  ],
  [
    0.9994891881942749,
    1.1696176443365403e-05,
    0.0004991634050384164
  ]
]
</pre>
<p>Success! Our TensorFlow model is 99.9% sure that this guy is a poodle!</p>
<h2>Next Steps</h2>
<p>There are obviously a lot of extensions required to turn this into a real-world production service. However, it does illustrate that hosting models as services is relatively straightforward and can be done quickly without writing a lot of boilerplate code.</p>
<p>There are a lot more components that go into turning this into a production-ready service and next time we&rsquo;ll look at how to set up a data pipeline for automatic retraining, version control of training data and more.</p>
<p class="attribution"><a href="http://bit.ly/2x5VLZD" class="ext-link" rel="external nofollow">Google</a> is a sponsor of The New Stack.</p>
<p class="attribution">Feature image <a href="http://ift.tt/2kGFWDT" class="ext-link" rel="external nofollow">via</a> Pixabay.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2zguXFZ">Poodle, Pug, or Wiener Dog? Deploy a Dog Identification TensorFlow Model Using Python and Flask</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
December 22, 2017 at 12:00PM
